### DISCUSSION TITLE
Nice idea:  Axmol 3.0 starts with "more" support of 3D. What are you mean?

### DISCUSSION THREAD

**Discussion by aismann** [axmolengine/axmol#1643] [General]
Maybe with some framework from here: #790 or written from scratch
Axmol 3.x.x is more the 3D line.
Axmol 2.x.x is more the 2D line.

**Comment by iAndyHD3** [comment]
3d nodes should be removed entirely. cocos2d-x was never a 3d engine, and as such 3d nodes are not used and they aren't mantained properly. It is a lot better to focus only on 2d, as that is the main use case of the engine. Removing 3d nodes will help to make the engine more lightweight, which is really needed at the moment IMO

**Reply by DelinWorks** [reply]
I disagree, having 3D support really opens up opportunities for advanced games not necessarily unreal engine level games. but just having access to an extra axis api allows developers to create clones similar minecraft or maybe use cool 3d effects in 2d with perspective cameras. basic 3D support should exist. SFML supports 2D only and I known they don't want to complicate the engine but if you wanna remove 3D from axmol might as well remove all extensions and libraries that are not used 100% of the time

**Comment by solan-solan** [comment]
Cocos2d-x itself supported 3d features always as I remember. And these features does not touch performance and all things have good architecture. Each popular engine has support of 2d and 3d equally.
I believe that it is because the main difference between these types of games is that the 3d require Mat4 and 2d - Mat3 calculations. 
It is true that there is a few people that use 3d, but why? I think that the main cause is absence of editor where you can build 3d scene and no understandable workflow how to create such games. 
About creating engine from scratch, it will require many time. Now even creator 3d have huge number of bugs after two years of release. 
Why not to think about how to add some 3d features to the existing engine and provide good workflow if there is really ideas to extend this part of engine? 
All 3d shaders, animation and 3d rendering works good as is now.

**Comment by aismann** [comment]
> ...  Why not to think about how to add some 3d features to the existing engine and provide good workflow if there is really ideas to extend this part of engine?

Thats is what I'm mean "from scratch". Not using another framework.
But why not using the data format from some of this frameworks and using the "free" editor also?

**Comment by solan-solan** [comment]
> But why not using the data format from some of this frameworks and using the "free" editor also?

I think it is good direction to extend 3d oportunities. But what do you mean under "data format from some of this frameworks"? Is this format of 3d models like fbx, glb/gltf?

About free editor the only thing that comes to mind is blender 3.8+, since it includes simple python scripting and wide possibilities for creating/placing models. Python script could generate some json file which include information about fbx/gltx models,materials(some material data like texture path), names of shaders. This json could be loaded as level file to the axmol (as example to inherited from ax::Scene class). Really such script could be done for 3dmax and many other proprietary programs, since it much simpler then doing separate editor.
Workflow:
1. create scene in blender
2. launch script which places all objects and level file to the Content axmol project folder.
3. launch axmol with appropriate level.

There are some flaw's: 
1. No opportunity to launch the scene from editor while developing
2. Missmatch between rendering of the scene in the blender/3dmax/ and axmol
3. No integrated physics simulating

Really it would be good for those who want to create some level with editor (for 3d platformer as example), but write other things using standard approach. May be some intermediate api which allow to find loaded from level file objects and endow them with additional properties like physics body, particles, animations, reacting to touch/keyboard

**Comment by rudiHammad** [comment]
Personally I would not go in a 3d direction with axmol because that would divert many resources in that area. I think focusing on 2d would make axmol incredibly solid as a 2d engine, specially considering that it's open source where the community helps on their free time.
In other words, try not to bit more that we can chew. But I might be wrong...
I also think, that even if axmol goes in a 3d direction, it would be fun at the beginning, but at the end users would keep expecting more and more, and probably end up going to unreal or unity. I don't know. Just my opinion.

**Comment by solan-solan** [comment]
Let me add one thing. I agree that imitation of unity/unreal would divert many resources, especially if you are fun of 2d. And I respect people that doing such games but there are some that prefer simple 3d games. Axmol includes skeletal animation/mesh rendering/mesh instancing/bullet physics and it is  quite enough to do some 3d application. 
What is game itself? It is gameplay firstly. But what unity/unreal can give from the box? Good graphics and thick framework. Other scripts/models/animation/...you should buy. In the best case you will get good looking picture on the screen. Is it enough to do game like genshin impact or something similar? I doubt. 
What gives axmol now. All needed to have simple 3d graphic. Other resources as well as gameplay should be implemented from the scratch like as with unity/unreal. I just wanted to note that it is absolutely equal for **indie** what he will use for 3d game (especially for mobile) axmol/unity/unreal, since he could not account to implement any big game (IMO). And in this case axmol looks preferable, since it is thin framework.

**Comment by aismann** [comment]
> ...but at the end users would keep expecting more and more, and probably end up going to unreal or unity. I don't know. Just my opinion.

- It's never planed to remove the 3D part or make an axmol 3D only version.
- Users/developer which want use more 3D or any other missing features on axmol, will add it (on his "axmol" fork) or go to another framework.

My opinion
As more 3D is on axmol (e.g. cpp-tests  examples) as more users will use axmol for some small 3D games.
